runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/CreateStreamTest.java

```diff
@@ -451,6 +457,47 @@ public void testAdvanceWatermarkEqualToPositiveInfinityThrows() {
     source.advanceWatermarkForNextBatch(BoundedWindow.TIMESTAMP_MAX_VALUE);
   }
 
+  @Test
+  public void testInStreamingModeCountByKey() throws Exception {
+    Instant instant = new Instant(0);
+
+    CreateStream<KV<Integer, Long>> kvSource =
+        CreateStream.of(KvCoder.of(VarIntCoder.of(), VarLongCoder.of()), batchDuration())
+            .emptyBatch()
+            .advanceWatermarkForNextBatch(instant)
+            .nextBatch(
+                TimestampedValue.of(KV.of(1, 100L), instant.plus(Duration.standardSeconds(3L))),
+                TimestampedValue.of(KV.of(1, 300L), instant.plus(Duration.standardSeconds(4L))))
+            .advanceWatermarkForNextBatch(instant.plus(Duration.standardSeconds(7L)))
+            .nextBatch(
+                TimestampedValue.of(KV.of(1, 400L), instant.plus(Duration.standardSeconds(8L))))
+            .advanceNextBatchWatermarkToInfinity();
+
+    PCollection<KV<Integer, Long>> output =
+        p.apply("create kv Source", kvSource)
+            .apply(
+                "window input",
+                Window.<KV<Integer, Long>>into(FixedWindows.of(Duration.standardSeconds(3L)))
+                    .withAllowedLateness(Duration.ZERO))
+            .apply(Count.perKey());
```

Comment:
you're lucky :) , Count is implemented in the sdk as Combine.PerKey which is translated in the current spark runner as GBK + Pardo and GBK's translation calls `SparkGroupAlsoByWindowViaWindowSet` which contains your fix. So your fix is called even if you call Count and not GBK. If the runner translated Combine.perKey directly (as the new spark runner) you would not be testing your fix.

Reply:
that is because translation in batch mode and in streaming mode are different: in streaming it uses the sdk impl based on GBK.